### PR TITLE
Tetu smart vault asset manager

### PR DESCRIPTION
**Pull request by AlehNat**

### CODE REVIEW

File: contracts/TetuVaultAssetManager.sol [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  // ***************************************************
+  //                VIEWS
+  // ***************************************************
+
+  /**
+   * @dev Checks balance of managed assets
+   */
+  function _getAUM() internal view override returns (uint256) {
+    return ISmartVault(tetuVault).underlyingBalanceInVault();
```
Comment (belbix) [inline contracts/TetuVaultAssetManager.sol:54]: underlyingBalanceWithInvestment is total balance

File: contracts/TetuVaultAssetManager.sol [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+   */
+  function _invest(uint256 amount) internal override returns (uint256) {
+    uint256 balance = underlying.balanceOf(address(this));
+    if (amount < balance) {
+      balance = amount;
+    }
+    uint256 sharesBefore = IERC20(tetuVault).balanceOf(address(this));
+
+    // invest to Tetu Vault
+    ISmartVault(tetuVault).depositAndInvest(balance);
```
Comment (belbix) [inline contracts/TetuVaultAssetManager.sol:74]: just deposit(), don't wast gas

File: contracts/TetuVaultAssetManager.sol (hunk 1 above)
Comment (belbix) [reply contracts/TetuVaultAssetManager.sol:54]: are you sure that you need total balance? maybe what owned by this contract?

File: contracts/TetuVaultAssetManager.sol [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  /**
+   * @dev Withdraws capital out of Tetu Vault
+   * @param amountUnderlying - the amount to withdraw
+   * @return the number of tokens to return to the balancerVault
+   */
+  function _divest(uint256 amountUnderlying) internal override returns (uint256) {
+    amountUnderlying = Math.min(amountUnderlying, IERC20(tetuVault).balanceOf(address(this)));
+    uint256 existingBalance = underlying.balanceOf(address(this));
+    if (amountUnderlying > 0) {
+      ISmartVault(tetuVault).withdraw(amountUnderlying);
```
Comment (belbix) [inline contracts/TetuVaultAssetManager.sol:91]: no, you need to use shares amount

File: contracts/TetuVaultAssetManager.sol [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return balance;
+  }
+
+  /**
+   * @dev Withdraws capital out of Tetu Vault
+   * @param amountUnderlying - the amount to withdraw
+   * @return the number of tokens to return to the balancerVault
+   */
+  function _divest(uint256 amountUnderlying) internal override returns (uint256) {
+    amountUnderlying = Math.min(amountUnderlying, IERC20(tetuVault).balanceOf(address(this)));
```
Comment (belbix) [inline contracts/TetuVaultAssetManager.sol:88]: you compare underlying with shares, is it correct?

**Review by belbix — CHANGES_REQUESTED** [review]

File: contracts/TetuVaultAssetManager.sol (hunk 3 above)
Comment (AlehNat) [reply contracts/TetuVaultAssetManager.sol]: you are correct. Fixed

File: contracts/TetuVaultAssetManager.sol (hunk 4 above)
Comment (AlehNat) [reply contracts/TetuVaultAssetManager.sol]: Fixed

File: contracts/TetuVaultAssetManager.sol (hunk 1 above)
Comment (AlehNat) [reply contracts/TetuVaultAssetManager.sol]: Correct. Fixed

File: contracts/TetuVaultAssetManager.sol (hunk 2 above)
Comment (AlehNat) [reply contracts/TetuVaultAssetManager.sol]: done

**Review by belbix — APPROVED** [review]

**Pull request merged**